RFC/RFC-14-p2p-adapter.md

```diff
@@ -0,0 +1,348 @@
+          if (route === 'server' || route.length > relayedRoute.length) {
+            route = relayedRoute
+          }
+        }
+      }
+    }
+
+    calculatedRoutes.set(toPeer, route)
+
+    // NOTE: routes are bidirectional
+    getOrCreateRoutingTable(toPeer).set(fromPeer, route === 'server' ? route : Array.from(route).reverse())
+    return route
+  }
+
+  const peers = new Set<string>()
+
+  for (const [peer, connections] of mesh) {
+    peers.add(peer)
+    for (const connection of connections) {
+      peers.add(connection)
+    }
+  }
+
+  for (const peerFrom of peers) {
+    for (const peerTo of peers) {
+      if (peerFrom === peerTo) {
+        continue
+      }
+
+      calculateRouteBeetwen(peerFrom, peerTo, [])
+    }
+  }
+
+  return routingTables
+}
+```
+
+## Packet
+
+A packet contains a source (peer id) and map specifying to whom the packet is for and the route to follow to reach it.
+
+```typescript
+
+type Packet = {
+  source: Address
+  to: Record<Address, Route>
+}
+```
+
+## Peer
+
+The peer will send a message to every known peer using the routing table, sending it once to each peer. Let's say peer1 is connected to peer2 and peer2 to peer3, when broadcasting to the network peer1 will send:
+
+```typescript
+{ 
+  source: 'peer1', 
+  to: { 'peer2': [], 'peer3': ['peer2'] }
+}
+```
+
+This means each peer will check the `to` field, if they are one of the recipent of the messages the message will be processed, otherwise it will be relayed in the way the route value indicates.
+
+Notice the peer relaying the message is not expected to remove itself either as a recipent or as a hop in the route, since this will require encoding the package again. 
+
+# Benefits
+
+- Since there is a specific routing table for each message, there is no need to expire message or count hops. If a route is cut, the message will not be delivered by the mesh, this means the message should be relayed using the messaging service (if reliable) or discarded (if unreliable).
+- The messaging service fallback provide a safety guarantee against network cluster partitions.
+- A given implementation can be optimized by suggesting peers to connect to certain others in order to avoid clustering and minimize messaging service usage. This is out of the scope for this document.
+- Since the routing service will know the status of the mesh at all times, it's easy to graph and debug network problems. 
+
+# Competition (alternatives)
+
+- One simple solution using a P2P mesh, in which peers broadcast to others, and upon receiving a message, they relay to their connections. The problem with this solution is there is the possibility of messages going thought the network after the same message is delivered, which implies the need to build a mechanism to discard old messages (using time or hops). Another problem is the possibility of clusters in the mesh, which may cause certain group of peers to not see the ones in another group (partitioned).
```

Comment:
```suggestion
- One simple solution using a P2P mesh, in which peers broadcast to others, and upon receiving a message, they relay to their connections. The problem with this solution is there is the possibility of messages going thought the network after the same message is delivered, which implies the need to build a mechanism to discard old messages (using time or hops). Another problem is the possibility of clusters in the mesh, which may cause certain group of peers to not see the ones in another group (partitioned).
- Keep the Routing Service to help building the Mesh, reducing the logic in peers and enabling lightweight ports of P2P comms to other technologies (using the same backend and mesh algorithms). Also use TURN to reduce the moving parts instead of the MessagingService
```

Reply:
We can continue this in another PR, I'm resolving the conversation in order to merge this one.